digest/src/core_api.rs

```diff
@@ -117,3 +117,15 @@ pub enum TruncSide {
     /// Truncate right side, i.e. `&out[m..]`.
     Right,
 }
+
+/// Core trait for saving internal state of the hash core and restoring it later.
+pub trait HashCoreInternalState {
```

Comment:
I'd suggest using Rust's built-in conversion traits rather than defining special methods like `internal_state` and `from_internal_state`:

```suggestion
pub trait HashCoreInternalState: TryFrom<Self::InternalState, Error = Error> + Into<Self::InternalState> {
```